### PR TITLE
Move stream decoding to StreamReader

### PR DESCRIPTION
**Pull request by sectioneight**
The Frame class was overloaded and rather smelly with all the
classmethods. This moves stream decoding into a StreamReader object
which uses a generator to more naturally iterate over incoming messages.

This also allows a single packet or chunk to contain multiple messages and have them all decoded and processed properly, which previously didn't work.

@uber/soap

### CODE REVIEW

File: tchannel/messages/call_request.py [hunk 1]
```diff
@@ -9,8 +9,19 @@ class CallRequestMessage(BaseMessage):
     message_type = Types.CALL_REQ
 
     __slots__ = (
+        'flags',
```
Comment (sectioneight) [inline tchannel/messages/call_request.py]: This is not really related to this PR. I should have put it in the parent PR since we now have all the fields defined for these message types.

File: tchannel/stream_reader.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+            if message_length > self.chunk_size:
+                rest_of_message = self._connection.read(
+                    message_length - self.chunk_size
+                )
+                if not rest_of_message:
+                    raise ProtocolException('Unexpectedly empty stream')
+
+                yield Frame.decode(BytesIO(chunk + rest_of_message))
+
+            elif len(chunk) > message_length:
```
Comment (sectioneight) [inline tchannel/stream_reader.py]: @abhinav you will probably point out that i don't need to guard this `if` condition and can let the final `else` case collapse into this. Do you think that's more readable? I think it makes the common case (message_size == size(chunk)) harder to understand but YMMV.

File: tchannel/stream_reader.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+from .exceptions import ProtocolException
+from .frame import Frame
+from .io import BytesIO
+from .parser import read_number
+from .parser import read_number_string
+
+
+class StreamReader(object):
+    """Read bytes from a stream and yield messages as they come."""
```
Comment (abhinav) [inline tchannel/stream_reader.py]: More documentation on this class would be nice. Especially things like,

* What's the type of connection? Is it a socket or file-like? 
* What is generated? Without looking at the implementation, I would have thought that it yields chunks of size `chunk_size`.
* What do the different initialization parameters mean?

File: tchannel/stream_reader.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
+from __future__ import absolute_import
+
+from .exceptions import ProtocolException
+from .frame import Frame
+from .io import BytesIO
+from .parser import read_number
+from .parser import read_number_string
+
+
+class StreamReader(object):
```
Comment (abhinav) [inline tchannel/stream_reader.py]: This should probably be called FrameReader or something similar. StreamReader suggests it has some relation to streams as defined in the `io` module.

File: tchannel/stream_reader.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+            elif len(chunk) > message_length:
+                # We got a payload with multiple messages.
+                stream = BytesIO(chunk)
+                remaining_bytes = len(chunk)
+
+                while remaining_bytes:
+                    message_length = read_number(stream, Frame.SIZE_WIDTH)
+                    yield Frame.decode(stream, message_length)
+                    remaining_bytes -= message_length
```
Comment (abhinav) [inline tchannel/stream_reader.py]: What happens if the chunk doesn't end with a message boundary? Like if the chunk ends right after you've read the message length?

Comment (sectioneight) [inline tchannel/stream_reader.py]: Yea that could be problematic. I don't think it would happen in normal use but we should be able to handle that.

**Comment by abhinav** [comment]
:shipit: 

I assume a separate diff will include the buffered reader.

**Comment by sectioneight** [comment]
I am going to do some more investigation to determine the best step going forward. In my own benchmarking, the performance characteristics are nearly exactly the same (~45K requests/second for a ping client with a single-threaded server and client) with the double-`recv` call. I think this is because the kernel buffers the incoming message for you.

tldr I may keep it this way since it seems to work well.
